packages/@uppy/provider-views/src/ProviderView/ProviderView.tsx

```diff
@@ -195,6 +244,194 @@ export default class ProviderView<M extends Meta, B extends Body> {
     }
   }
 
+  #hasServerSideSearch(): boolean {
+    const supportsServerSearch = typeof this.provider.search === 'function'
+    return supportsServerSearch
+  }
+
+  async #performSearch(): Promise<void> {
+    const { partialTree, currentFolderId, searchString } =
+      this.plugin.getPluginState()
+    const currentFolder = partialTree.find(
+      (i) => i.id === currentFolderId,
+    ) as PartialTreeFolder
+
+    this.setLoading('Searching...')
+    await this.#withAbort(async (signal) => {
+      const scopePath = currentFolder.type === 'root' ? null : currentFolderId
+      const { items } = await (this.provider as any).search(searchString, {
+        signal,
+        path: scopePath ?? undefined,
+      })
+
+      // ! any alternatives ?
+      // return if user clears search before results arrive
+      const currentSearchString = this.plugin.getPluginState().searchString
+      if (currentSearchString.trim() === '') {
+        return
+      }
+
+      this.#searchState.searchResult = items
+      this.#searchState.scopeId = scopePath
+      this.plugin.setPluginState({ isSearchActive: true })
+    }).catch(handleError(this.plugin.uppy))
+    this.setLoading(false)
+  }
+
+  onSearchInput = (s: string): void => {
+    this.plugin.setPluginState({ searchString: s })
+    this.#clearSearchDebounce()
+    const trimmed = s.trim()
+
+    if (trimmed === '') {
+      this.clearSearchState()
+      return
+    }
+
+    if (!this.#hasServerSideSearch()) {
+      return
+    }
+
+    this.#searchState.debounceId = window.setTimeout(() => {
+      this.#performSearch()
+      this.#searchState.debounceId = undefined
+    }, 500)
+  }
+
+  // Create a minimal CompanionFile object for ancestor folders when building the path
+  #createMinimalFolderData(name: string, requestPath: string): CompanionFile {
+    return {
+      id: `synthetic:${requestPath}`,
+      name: decodeURIComponent(name),
+      icon: 'folder',
+      type: 'folder',
+      mimeType: '',
+      extension: '',
+      size: 0,
+      isFolder: true,
+      modifiedDate: '',
+      thumbnail: '',
+      requestPath,
+    }
+  }
+
+  // build the Leaf Node , it can be a file ( PartialTreeFile ) or a folder ( PartialTreeFolderNode )
+  // Since we Already have the leaf node's data ( file : CompanionFile) from the searchResults: CompanionFile[]  , we just use that.
+  #buildLastNode(
```

Comment:
Same here, can you inline this into `buildPath`, separating it out here only makes you think it's used elsewhere too but it isn't, it's tightly coupled to the search functionality.